packages/js/src/plugins/auctionHouseModule/operations/buy.ts

```diff
@@ -0,0 +1,214 @@
+import { ConfirmOptions, PublicKey } from '@solana/web3.js';
+import type { Metaplex } from '@/Metaplex';
+import { Option } from '@/utils';
+import {
+  Operation,
+  OperationHandler,
+  Pda,
+  Signer,
+  SolAmount,
+  SplTokenAmount,
+  useOperation,
+} from '@/types';
+import { SendAndConfirmTransactionResponse } from '../../rpcModule';
+import { AuctionHouse, Listing, Purchase } from '../models';
+
+// -----------------
+// Operation
+// -----------------
+
+const Key = 'BuyOperation' as const;
+
+/**
+ * Creates a bid on a given asset and then executes a sale on the created bid and listing.
+ *
+ * ```ts
+ * await metaplex
+ *   .auctionHouse()
+ *   .buy({ auctionHouse, listing, buyer })
+ *   .run();
+ * ```
+ *
+ * @group Operations
+ * @category Constructors
+ */
+export const buyOperation = useOperation<BuyOperation>(Key);
+
+/**
+ * @group Operations
+ * @category Types
+ */
+export type BuyOperation = Operation<typeof Key, BuyInput, BuyOutput>;
+
+/**
+ * @group Operations
+ * @category Inputs
+ */
+export type BuyInput = {
+  /** The Auction House in which to create a Bid and execute a Sale. */
+  auctionHouse: AuctionHouse;
+
+  /**
+   * The Auction House authority.
+   * If this is Signer the transaction fee
+   * will be paid from the Auction House Fee Account
+   *
+   * @defaultValue `auctionHouse.authority`
+   */
+  authority?: PublicKey | Signer;
+
+  /**
+   * Creator of a bid.
+   *
+   * @defaultValue `metaplex.identity()`
+   */
+  buyer?: Signer;
+
+  /**
+   * The Listing that is used in the sale.
+   * We only need a subset of the `Listing` model but we
+   * need enough information regarding its settings to know how
+   * to execute the sale.
+   *
+   * This includes, its asset, auction house address, seller, receipt address etc.
+   */
+  listing: Pick<
+    Listing,
+    | 'asset'
+    | 'auctionHouse'
+    | 'canceledAt'
+    | 'sellerAddress'
+    | 'tradeStateAddress'
+    | 'receiptAddress'
+  >;
+
+  /**
+   * The Auctioneer authority key.
+   * It is required when Auction House has Auctioneer enabled.
+   *
+   * @defaultValue No default value.
+   */
+  auctioneerAuthority?: Signer;
+
+  /**
+   * The token account address that's associated to the asset a bid created is for.
+   * If this or seller isn't provided in the Listing, then the bid will be public.
+   *
+   * @defaultValue No default value.
+   */
+  tokenAccount?: Option<PublicKey>;
```

Comment:
This is not necessary since we're buying the listing's asset directly. It will always be a private bid and we know the token account via the `listing` parameter.